### PR TITLE
 IUC want profile in the <tool> and detect_errors in the <command>

### PR DESCRIPTION
**Pull request by fubar2**
Hi @hexylena! 
Request from the IUC to add boilerplate doodads to ToolFactory outputs, so this PR elevates command to an XMLParam, adding the command string as a CDATA text. It passes tests including the new one and more importantly, the existing one, though I'm not sure if this is the most efficient path. Whatever, it works for me FWIW - thanks!

### CODE REVIEW

**Review by bernt-matthias — COMMENTED** [review]
Good idea to allow setting a tool's `profile` and a command's `detect_errors` property. 
Is setting `detect_errors`  to anything else than `"aggressive"` supported?
Can you add a test showing how to set non-default values?

**Comment by hexylena** [comment]
>  Is setting detect_errors to anything else than "aggressive" supported?


<del>yes, and aggressive is generally the wrong choice, IMO. I would much, much rather see it set to `exit_code`</del>

<del>Aggressive is rather annoying when wrapping other people's tools; as soon as some library decides to log a non-fatal warning or error, the job fails, despite the job in fact succeeding by exit code/results. Too many tools have used "Error:" in their outputs that I avoid aggressive.</del>

edit: you meant in the python side. Yes, would like a test for that. (And a diffrent default)

**Comment by hexylena** [comment]
@bernt-matthias where are you with #37, should we merge it? I'd like the github workflow that you've done there.

**Comment by bernt-matthias** [comment]
> @bernt-matthias where are you with #37, should we merge it? I'd like the github workflow that you've done there.

Ready for review... Not sure if tests succeed yet. Can also separate the workflow in a separate PR if needed.

**Comment by fubar2** [comment]
Thanks @hexylena  and @bernt-matthias!

Yes, defaults are both what @bgruening asked for in https://github.com/bgruening/galaxytools/pull/1326. AFAIK, those should properly be determined by the IUC I think, not me - I'm just doing what I was asked and will be happy to meet their requirements.
```
[bgruening](https://github.com/bgruening) [Sep 1, 2023](https://github.com/bgruening/galaxytools/pull/1326#discussion_r1312980881)

can we please add profile="22.05"
```
and 
```
[bgruening](https://github.com/bgruening) [Sep 1, 2023](https://github.com/bgruening/galaxytools/pull/1326#discussion_r1312983046)

This very simple stdio is deprecated and should be replaced with <command detect_errors="aggressive">
```
They should both be explicitly set by the caller, like this example from the ToolFactory for profile:
```
        self.newtool = gxt.Tool(
            self.tool_name,
            self.tool_id,
            self.tool_version,
            self.args.tool_desc,
            self.args.sysexe,
            profile = self.profile
```
 and here for detect_errors:
`
self.newtool.command = gxtp.Command(detect_errors="aggressive")`

These changes already have one new unit test and import_xml_sample has these elements - and they appear to be correctly parsed and set in the unit test. The ToolFactory tests them too if I use this updated galaxyxml and it passes its own tests. 
In summary: I am not sure exactly how to satisfy this beyond the already added unit test - hints appreciated. Are you suggesting another unit test @bernt-matthias? One to set and test different values. I could do that if you and @hexylena think it would help get this sorted out.

The defaults are the settings suggested. Perhaps too conservative and obviously the caller has the power to set them as they wish. Happy to change them if someone who is better placed to advise than Bjoern is willing to make a ruling?. 
_Please let me know explicitly what I need to do to get this in so I can get on using it?_

**Comment by fubar2** [comment]
@hexylena @bernt-matthias 
Does this additional unit test help? 
The constructor is being called from import_xml.py (python side) in the unit tests and seems to work. Do you want something outside the unit tests to do that too? So these new properties are already being read and stowed from sample xml and seem just dandy in that first new unit test, but another explicitly setting and testing the new values, confirm that they are mutable, so can be whatever the user wants as the ToolFactory constructor examples show. 

Is anything not yet being tested that needs testing for these changes to be acceptable?
 Please let me know what else I need to do?

```
class TestChangeProfDet(TestImport):
    def test_changes(self):
        self.tool.command.node.attrib["detect_errors"] = "foobarfoo"
        de = self.tool.command.node.attrib["detect_errors"]
        self.assertEqual(de, "foobarfoo")
        self.tool.profile  = "anything you want"
        self.assertEqual(self.tool.profile,  "anything you want")
```

**Comment by bernt-matthias** [comment]
I guess the most undebatable option is to use Galaxy's defaults, ie to not set them by default.

Btw would this be done by setting the value to `None`? 

But I'm fine with any default.

**Comment by fubar2** [comment]
Thanks @bernt-matthias - So all good? 
I no longer see real use so I have no experience with their effects, so changing them to anything else is fine with me. 
Default being None makes them required at init I guess which might be wise if they are critically important, like an id.
Strict defaults for anyone not making their own choice seems better than making them required to me FWIW. 
ToolFactory generated tools will have those values because that's what @bgruening asked for.

**Comment by bernt-matthias** [comment]
For your PRs I would suggest to go with marius' suggestion https://github.com/bgruening/galaxytools/pull/1326#discussion_r1314559091, ie set `profile` and no `detect_errors`

**Comment by fubar2** [comment]
ok. one moment please....

**Comment by bernt-matthias** [comment]
I don't see in your coffee the possibility to omit the two attributes from the generated xml

**Comment by fubar2** [comment]
ok Thanks again @bernt-matthias.
Default is None for detect_errors now.
import_xml.xml now has profile and no detect_errors.
It parses with testimport.py as expected giving 22.05 profile
and an empty command tag now.
Unit tests all pass with tox.


```
ross@pn50:~/rossgit/galaxyxml/test$ python testimport.py import_xml.xml
reading from import_xml.xml             
INFO:galaxyxml.tool.import_xml:<description> is loaded during initiation of the object.
INFO:galaxyxml.tool.import_xml:<stdio> loaded.                     
INFO:galaxyxml.tool.import_xml:<version_command> is loaded during initiation of the object.
<tool name="Import" id="import_test" version="1.0" profile="22.05">
  <description>description</description>                                                  
  <edam_operations>                      
    <edam_operation>operation_0004</edam_operation>                     
  </edam_operations>
  <edam_topics>
    <edam_topic>topic_0003</edam_topic>
  </edam_topics>                                                                          
  <requirements>     
    <requirement version="1" type="package">magic_package</requirement>
    <container type="docker">a_container</container>
  </requirements>
  <stdio>                                                                                 
    <exit_code range="1:" level="fatal"/>
  </stdio> 
  <version_command><![CDATA[v_command]]></version_command> 
  <command><![CDATA[                                                                                                                                                                
command interval_file $interval_file
$bool       
int_size $int_size
float_size $float_size
```

**Comment by fubar2** [comment]
>I don't see in your coffee the possibility to omit the two attributes from the generated xml

Thanks again, @bernt-matthias but if you remove profile from the sample, it parses without a problem, and detect errors is also not there or in the output - so missing in the input is harmless. However, the emitted xml has what was supplied or the default profile if there was not one, satisfying https://galaxy-iuc-standards.readthedocs.io/en/latest/best_practices/tool_xml.html?highlight=profile#tool-profile and the rule that one should be forgiving of what is ingested but strict in what is emitted.

Are there situations where that guideline should not be followed in generating XML? If we are not following those guidelines, it's going to be hard to make a good decision.

**Comment by fubar2** [comment]
@hexylena @bernt-matthias: please let me know what outstanding issues remain that block this from being merged?

File: galaxyxml/tool/import_xml.py [hunk 1]
```diff
@@ -25,12 +25,10 @@ def _init_tool(self, xml_root):
         for child in xml_root:
             if child.tag == "description":
                 description = child.text
-            elif child.tag == "command":
-                executable = child.text.split()[0]
-                command = child.text
```
Comment (bernt-matthias) [inline galaxyxml/tool/import_xml.py:30]: Why is command removed here?

File: test/unit_test_import_xml.py [hunk 2]
```diff
@@ -27,15 +26,34 @@ def test_override(self):
...
         exml = exml.replace("\n", " ")
+        print("-----------exml", exml)
         self.assertTrue(co in exml)
 
+class TestCommand(TestImport):
+    def test_command(self):
+        try:
+            de = self.tool.command.node.attrib["detect_errors"]
+        except KeyError:
+                de = None
```
Comment (bernt-matthias) [inline test/unit_test_import_xml.py]: Fix indentation here.

File: test/unit_test_import_xml.py [hunk 3]
```diff
@@ -27,15 +26,34 @@ def test_override(self):
...
+        self.assertEqual(de, None)
+        ctext = self.tool.command.node.text
+        self.assertEqual(ctext.strip(), "command")
+
+class TestChangeProfDet(TestImport):
+    def test_changes(self):
+        self.tool.command.node.attrib["detect_errors"] = "foobarfoo"
+        de = self.tool.command.node.attrib["detect_errors"]
+        self.assertEqual(de, "foobarfoo")
```
Comment (bernt-matthias) [inline test/unit_test_import_xml.py]: I don't understand this test. Looks like:

```
A = "foobarfoo"
B = A
self.assertEqual(B, "foobarfoo")
```

Same for the following two lines.

File: test/import_xml.xml [hunk 4]
```diff
@@ -1,5 +1,5 @@
 <?xml version="1.0"?>
-<tool id="import_test" name="Import" version="1.0">
```
Comment (bernt-matthias) [inline test/import_xml.xml]: The xml examplifies the defaults, or?

**Comment by hexylena** [comment]
> However, the emitted xml has what was supplied or the default profile if there was not one, satisfying https://galaxy-iuc-standards.readthedocs.io/en/latest/best_practices/tool_xml.html?highlight=profile#tool-profile and the rule that one should be forgiving of what is ingested but strict in what is emitted.

This library isn't intended to generate IUC best practice tools, that's not it's goal, it's intended to let you flexibly generate the tool you need. 

> Are there situations where that guideline should not be followed in generating XML? If we are not following those guidelines, it's going to be hard to make a good decision.

this is a prescriptivist take (these are our rules and they are the only right way) and this library is targetted at people that know what they want to do and want flexibility, not enforcing IUC best practices (as you'll note, it's not under galaxy-iuc)

File: galaxyxml/tool/__init__.py [hunk 5]
```diff
@@ -134,49 +131,52 @@ def export(self, keep_old_command=False):
...
+                export_xml.append(stdio_element)
+            except Exception:
+                export_xml.append(Expand(macro="stdio"))
 
         # Append version command
         export_xml.append_version_command()
 
         if self.command_override:
-            command_line = self.command_override
+            command_text = self.command_override
```
Comment (hexylena) [inline galaxyxml/tool/__init__.py]: why did you rename this?

File: galaxyxml/tool/import_xml.py [hunk 6]
```diff
@@ -77,21 +74,26 @@ def _load_stdio(self, tool, stdio_root):
...
-    def _load_command(self, tool, desc_root):
+    def _load_command(self, tool, command_root):
         """
-        <command> is already loaded during initiation.
-
-        :param tool: Tool object from galaxyxml.
-        :type tool: :class:`galaxyxml.tool.Tool`
-        :param desc_root: root of <command> tag.
-        :type desc_root: :class:`xml.etree._Element`
+        now an XMLParameter with a text
```
Comment (hexylena) [inline galaxyxml/tool/import_xml.py]: please don't delete documentation, update it if it needs updating.

File: galaxyxml/__init__.py [hunk 7]
```diff
@@ -1,7 +1,4 @@
-from builtins import (
-    object,
-    str
-)
+from builtins import object, str
```
Comment (hexylena) [inline galaxyxml/__init__.py:1]: yes we should have been running `make reformat` more regularly, but in the future please don't include reformatting changes mixed in with your PR, it makes it much more difficult to review. Reformatting should be a separate PR.

File: galaxyxml/tool/__init__.py [hunk 8]
```diff
@@ -134,49 +131,52 @@ def export(self, keep_old_command=False):
...
                 raise
             try:
-                command_line.append(export_xml.outputs.cli())
+                command_text.append(export_xml.outputs.cli())
             except Exception:
                 pass
-        # Steal interpreter from kwargs
-        command_kwargs = {}
-        if export_xml.interpreter is not None:
-            command_kwargs["interpreter"] = export_xml.interpreter
```
Comment (hexylena) [inline galaxyxml/tool/__init__.py:164]: yes, ok, this should not have been included. good. https://docs.galaxyproject.org/en/latest/dev/schema.html#id22 @bernt-matthias the documentation looks a bit odd there.

File: galaxyxml/tool/__init__.py [hunk 9]
```diff
@@ -134,49 +131,52 @@ def export(self, keep_old_command=False):
...
-        # Add command section
-        command_node = etree.SubElement(export_xml.root, "command", **command_kwargs)
         if keep_old_command:
-            if getattr(self, "command", None):
-                command_node.text = etree.CDATA(export_xml.command)
+            if getattr(self, "command_line", None):
+                ctext = export_xml.command_text
```
Comment (hexylena) [inline galaxyxml/tool/__init__.py]: now that you've renamed command_line, this should be command_text, right?

```suggestion
            if getattr(self, "command_text", None):
                ctext = export_xml.command_text
```

**Review by bernt-matthias — COMMENTED** [review]
I think there are now some diffs in the example:

```
        python examples/example.py > tmp.xml
        xmldiff tmp.xml examples/tool.xml
        python examples/example_macros.py > tmp.xml
        xmldiff tmp.xml examples/example_macros.xml
```

Can you fix those?

File: galaxyxml/tool/__init__.py [hunk 10]
```diff
@@ -134,49 +131,52 @@ def export(self, keep_old_command=False):
...
-                actual_cli = "%s %s" % (export_xml.executable, export_xml.clean_command_string(command_line),)
-            command_node.text = etree.CDATA(actual_cli.strip())
-        export_xml.append(command_node)
+                actual_cli = "%s %s" % (
+                    export_xml.executable,
+                    export_xml.clean_command_string(command_text),
+                )
+            ctext = actual_cli.strip()
+        export_xml.command_text = ctext
+        ctext = '\n' + ctext # pretty - bjoern's suggestion
```
Comment (hexylena) [inline galaxyxml/tool/__init__.py]: I don't love this tbh, it'll introduce lots of diffs to add some useless whitespace. If you want a newline in your command add it, if you want it pretty, run it through planemo. This isn't the job of this library

File: tox.ini [hunk 11]
```diff
@@ -1,3 +1,15 @@
...
 max-line-length = 230
 max-complexity = 10
+[testenv]
+description = run the tests with pytest
+package = wheel
+wheel_build_env = .pkg
+deps =
+    pytest>=6
+commands =
+    pytest test/unit_test_import_xml.py {tty:--color=yes} {posargs}
```
Comment (hexylena) [inline tox.ini]: We have not been using tox to test, this is a bigger change, hmm. until now it was all `nose`, which apparently is [now broken](https://github.com/nose-devs/nose/issues/1099). I _really_ wish some of these changes had come in separate PRs so we could've discussed them in isolation, but I guess we have to accept this because nose is non functional.

**Comment by hexylena** [comment]
@bernt-matthias you have most of the same review comments I did :sweat_smile: I should've hit submit earlier

File: galaxyxml/tool/import_xml.py (hunk 1 above)
Comment (hexylena) [reply galaxyxml/tool/import_xml.py:30]: it took me a while to understand this, it's because now the command is parsed in `Command`, rather than expecting the entire thing to be set from the start. I was also unhappy it was re-ordered since it made the diff unnecessarily confusing.

**Review by hexylena — COMMENTED** [review]
As of now it doesn't pass the tests (the other tests, in the .travis.yml that we didn't finish turning into a GHA.). I will update those @fubar2 .

```diff
$ diff tmp.xml examples/tool.xml 
1c1
< <tool name="aragorn" id="se.lu.mbioekol.mbio-serv2.aragorn" version="1.2.36" profile="22.05">
---
> <tool name="aragorn" id="se.lu.mbioekol.mbio-serv2.aragorn" version="1.2.36">
7a8,11
>   <stdio>
>     <exit_code range="1:" level="fatal"/>
>   </stdio>
>   <expand macro="stdio"/>
9,10c13
<   <command><![CDATA[
< aragorn.exe $flag
---
>   <command><![CDATA[aragorn.exe $flag
```

and yeah, fine with dropping the stdio, but still not crazy about adding profile by default, that really seems like something for library consumers to do. I also don't understand the newline being added there.

**Comment by hexylena** [comment]
I would recommand also that don't make PRs from your `master` branch in the future, it's best to make them from a feature branch.

**Comment by hexylena** [comment]
@fubar2 @bernt-matthias 

```
        python examples/example.py > tmp.xml
        xmldiff tmp.xml examples/tool.xml
        python examples/example_macros.py > tmp.xml
        xmldiff tmp.xml examples/example_macros.xml
```

I fixed the first set, I'm still struggling to figure out why the second set changed, @fubar2 maybe you can fix that.

**Comment by fubar2** [comment]
ok - thanks for the comments - appreciated.
 
Most problems are from changes to make `<command>`  an XMLParam. 
Please let me know if you really want to leave them in this PR - would prefer to revert them otherwise.

Apologies for the import sample changes - from using it to confirm that either missing or supplied values were handled - shouldn't have committed those.
Yes, too much at once and yes, a branch next time after this reminder...
I'll try to figure the rest out and make adjustments in response after I revert the command stuff?

File: test/unit_test_import_xml.py (hunk 3 above)
Comment (fubar2) [reply test/unit_test_import_xml.py]: yes. Makes sure they are persistent and mutable after the previous test checks the correct values have been extracted from the sample. Was trying to respond to a previous suggestion. Will remove it.

File: galaxyxml/tool/__init__.py (hunk 5 above)
Comment (fubar2) [reply galaxyxml/tool/__init__.py]: Plan to revert all this - part of the command attributes change.
Took me a while to get my head around where the actual command text was being found in the new XMLParam and I discovered I was reusing a variable name. Went the wrong way - sorry - should have backed out but forged ahead with new names. Will soon be back to what it was.

Comment (fubar2) [reply galaxyxml/tool/__init__.py]: The stdio default insertion is dropped. 
Stdios still work fine if you want to use them. :) 
This all started when I was asked to add profile and remove that deprecated default stdio value, in a review of some tools generated using galaxyxml, so not sure what else I can do here.

**Comment by hexylena** [comment]
It happens, no worries. Let's not dwell.

- Please feel free to update this PR accordingly, I've already pushed some changes to this branch :) 
- Making it a proper `Command` is a good change and one that I would like to see kept, as I see it that's one of the core improvements in this PR.
- If the tests pass, well, that's what's really important in the end.

**Comment by fubar2** [comment]
ok. 
Thanks again!

File: galaxyxml/tool/import_xml.py (hunk 1 above)
Comment (fubar2) [reply galaxyxml/tool/import_xml.py:30]: not a deliberate or useful reorder - now less reordered.

File: galaxyxml/__init__.py (hunk 7 above)
Comment (fubar2) [reply galaxyxml/__init__.py:1]: Not intentional just can't figure out how to use tox test without being reformatted :( sorry.

**Comment by fubar2** [comment]
> > However, the emitted xml has what was supplied or the default profile if there was not one, satisfying https://galaxy-iuc-standards.readthedocs.io/en/latest/best_practices/tool_xml.html?highlight=profile#tool-profile and the rule that one should be forgiving of what is ingested but strict in what is emitted.
> 
> This library isn't intended to generate IUC best practice tools, that's not it's goal, it's intended to let you flexibly generate the tool you need.
> 
> > Are there situations where that guideline should not be followed in generating XML? If we are not following those guidelines, it's going to be hard to make a good decision.
> 
> this is a prescriptivist take (these are our rules and they are the only right way) and this library is targetted at people that know what they want to do and want flexibility, not enforcing IUC best practices (as you'll note, it's not under galaxy-iuc)

Apologies - my immediate problem is to get things past review elsewhere - sorry it spilled over where it does not belong.  I'll  advocate for the changed defaults recommended by those reviewers, but all those decisions are entirely yours...

File: galaxyxml/__init__.py (hunk 7 above)
Comment (fubar2) [reply galaxyxml/__init__.py:1]: ok - sorry! Unintended changes sjhould now be minimal, but tox...

**Comment by fubar2** [comment]
I hope all the worst is gone and it's ready for another review. 
Reformatting is from running tests in the dumbest possible way so not intended but effective.
Sorry for the noise. 
Defaults for profile and command detect_errors are None.
The IUC guidance seems way out of date and hard to know which bits to take seriously. Disappointing and hard to support really. A condign punishment.

**Comment by hexylena** [comment]
>  The IUC guidance seems way out of date and hard to know which bits to take seriously. Disappointing and hard to support really. A condign punishment.

It would be great to get it updated but, things move very very slowly there it seems. c.f. my indentation PR allowing tabs, which is now dormant since months. We also didn't have the 'traditional' IUC meeting at GCC too, so, didn't have the usual going through of issues/etc.

**Review by hexylena — APPROVED** [review]
This round was much better to review, thanks for changes @fubar2.

**Comment by fubar2** [comment]
github action flake now unhappy so format issues to resolve or silence?

**Comment by hexylena** [comment]
Feel free to ignore flake. I've pushed a commit to take care of one test, then there's only one more failure:

```
12c12,13
<   <token name="ARAGORN_OUTMACRO"><![CDATA[]]></token>
---
>   <token name="ARAGORN_OUTMACRO"><![CDATA[-output '$output'
> ## TODO CLI for OutputCollection collection]]></token>
```

Previously the outmacro had included an `-output '$output'` with a TODO for the collection, now that seems to be missing, but I think this was an issue already for a while. You contributed the collection support ages ago (awesome thank you :yellow_heart:) but I think our tests haven't been working since then so we didn't notice until now.

So I will merge over this specific failure, and maybe if you have energy and bandwidth, if you have a nice fix for it, i think it would be useful to the library's consumers to retain this todo/-output

<details><summary>flake rant</summary>
 It's stuff black will take care of which is especially, especially frustrating for me. Why lint and make a human fix it, just fix it and shut up (I say to flake/black.) These ridiculous tools should either fix it automatically because it's bloody whitespace, or stop complaining. `isort` is banned from all of my projects because it drives me up a wall.
</details>

**Review by hexylena — APPROVED** [review]

**Comment by hexylena** [comment]
Thanks for the patience @fubar2 and putting up with our complaints. I'll mint a new release for ya

**Comment by hexylena** [comment]
after a silly amoutn of fighting with pypi/gha, https://pypi.org/project/galaxyxml/0.5.3/ now exists.
